docs/architecture/adr-004-split-denomination-keys.md

```diff
@@ -0,0 +1,58 @@
+# ADR 004: Split Denomination Keys
+
+## Changelog
+
+- 2020-01-08: Initial version
+- 2020-01-09: Alterations to handle vesting accounts
+
+## Context
+
+> This section describes the forces at play, including technological, political, social, and project local. These forces are probably in tension, and should be called out as such. The language in this section is value-neutral. It is simply describing facts.
+> {context body}
+
+## Decision
+
+Balances shall be stored per-account & per-denomination under a denomination- and account-unique key, thus enabling O(1) read & write access to the balance of a particular account in a particular denomination.
+
+### Account interface (x/auth)
+
+`GetCoins() and `SetCoins()` will be removed from the account interface, since coin balances will now be stored in & managed by the bank module.
+
+`GetVestedCoins()`, `GetVestingCoins()`, `GetOriginalVesting()`, `GetDelegatedFree()`, and `GetDelegatedVesting()` will be altered to take a bank keeper and a denomination as two additional arguments, which will be used to lookup the balances from the base account as necessary.
+
+Vesting accounts will continue to store original vesting, delegated free, and delegated vesting coins (which is safe since these cannot contain arbitrary denominations).
+
+### Bank keeper (x/bank)
+
```

Comment:
Maybe add a short section on the key structure?

Reply:
Aye, added.